File: 03-java-socket/src/main/java/demo/HelloProtelis.java
package demo;

import com.uchuhimo.konf.BaseConfig;
import com.uchuhimo.konf.Config;
import demo.data.ProtelisNode;
import org.protelis.lang.ProtelisLoader;
import org.protelis.vm.ProtelisProgram;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * Example usage of the implemented classes.
 */
public final  class HelloProtelis {

    private HelloProtelis() { }

    /**
     * Main method.
     * @param args unutilized
     */
    public static void main(final String[] args) {
        Config config = new BaseConfig();
        config.addSpec(ProtelisConfigSpec.SPEC);
        config = config.from().toml.resource("config.toml");
        final String protelisModuleName = config.get(ProtelisConfigSpec.protelisModuleName);
        final int iterations = config.get(ProtelisConfigSpec.iterations);
        final List<ProtelisNode> nodes = config.get(ProtelisConfigSpec.nodes);
        final List<Device> devices = new ArrayList<>();
        nodes.forEach(n -> {
            final SocketNetworkManager netmgr = new SocketNetworkManager(new IntDeviceUID(n.getId()), n.getHostandport().getPort(), n.getNeighbors());
            try {
                netmgr.listen();
            } catch (IOException e) {
                e.printStackTrace();
            }
            final ProtelisProgram program = ProtelisLoader.parse(protelisModuleName);
            final Device node = new Device(program, n.getId(), netmgr, new ConsoleSpeaker());
            if (n.isLeader()) {
                node.getDeviceCapabilities().getExecutionEnvironment().put("leader", true);
            }
            devices.add(node);
        });
        // Run some cycles
        for (int i = 0; i < iterations; i++) {
            devices.forEach(Device::runCycle);
        }
        // Stop the server socket
        devices.forEach(d -> ((SocketNetworkManager) d.getNetworkManager()).stop());
    }
}

File: 04-kotlin-socket/src/main/kotlin/HelloProtelis.kt
package demo

import com.uchuhimo.konf.Config
import org.protelis.lang.ProtelisLoader

class Main {

    private var devices: List<Device> = emptyList()
    private val config = Config { addSpec(ProtelisConfigSpec) }
            .from.toml.resource("config.toml")
    private val protelisModuleName = config[ProtelisConfigSpec.protelisModuleName]
    private val iterations = config[ProtelisConfigSpec.iterations]
    private val nodes = config[ProtelisConfigSpec.nodes]

    fun hello() {
        // Initialize some nodes.
        nodes.forEach {
            val socketNetworkManager = SocketNetworkManager(IntDeviceUID(it.id), it.hostandport.port, it.neighbors).apply { listen() }
            val program = ProtelisLoader.parse(protelisModuleName)
            val node = Device(program, it.id, socketNetworkManager)
            if (it.leader) {
                node.deviceCapabilities.executionEnvironment.put("leader", true)
            }
            devices += node
        }
        // Let the nodes make some iterations.
        repeat(iterations) {
            devices.forEach { it.runCycle() }
            devices.forEach { it.sendMessages() }
        }
        // Close the thread listening.
        devices.forEach { (it.netmgr as SocketNetworkManager).stop() }
    }
}

fun main() {
    Main().hello()
}

